### ISSUE TITLE
Error in authorization for Manual DNS verification

### ISSUE THREAD

**Issue by chandan1001** [rmbolger/Posh-ACME#103]
I am trying to generate DNS verified certificates.

 I followed [https://github.com/rmbolger/Posh-ACME/wiki/%28Advanced%29-Semi-Manual-DNS-Challenge-Validation#notify-the-acme-server](url)

Steps I followed are -

```

1. Import-Module posh-acme
 
2. Set-PAServer LE_STAGE
3. New-PAAccount -AcceptTOS -Contact 'me@mymail.com'
4. New-PAOrder 'devcorp.mydomain.com','*.devcorp.mydomain.com'

5. $auths = Get-PAOrder | Get-PAAuthorizations
 
6. $auths.DNS01Url | Send-ChallengeAck

```

```
(Get-PAOrder | Get-PAAuthorizations).DNS01TOKEN
```
Executing above command after line 5. give below tokens

```
uzPAA3Xd6ZaS9XcNaV06JCD-dTn9TqDf2iga_XKjeTs
90JYwKv0JI4TGaXh5jspUwqFK0y-oDO6krrSIEaFQeI
```

After line 6 
```
(Get-PAOrder | Get-PAAuthorizations).challenges.error | fl

```
gives

```
type   : urn:ietf:params:acme:error:unauthorized
detail : Incorrect TXT record "uzPAA3Xd6ZaS9XcNaV06JCD-dTn9TqDf2iga_XKjeTs" (and 1 more) found at _acme-challenge.devcorp.mydomain.com
status : 403

type   : urn:ietf:params:acme:error:unauthorized
detail : Incorrect TXT record "90JYwKv0JI4TGaXh5jspUwqFK0y-oDO6krrSIEaFQeI" (and 1 more) found at _acme-challenge.devcorp.mydomain.com
status : 403


```

Where am I doing it wrong ?

**Comment by rmbolger** [comment]
I believe you missed a small part of the next step in [Publishing a DNS Challenge](https://github.com/rmbolger/Posh-ACME/wiki/%28Advanced%29-Semi-Manual-DNS-Challenge-Validation#publishing-a-dns-challenge).

The `DNS01Token` values don't get created as-is, they need to be converted into "key authorization" values which are the token plus the account's public key thumbprint. If you pass the `DNS01Token` values as an argument to `Get-KeyAuthorization`, you should get the actual values that need to go into the TXT record.

**Comment by chandan1001** [comment]
Thanks..

I tried your suggestion 

```
PS C:\WINDOWS\system32> (Get-PAOrder | Get-PAAuthorizations).DNS01TOKEN |Get-KeyAuthorization
uBFahyhvNjrhkVEJcruLq0wssSnowmNWa12jYETts3c.zqngmHPWrG6hTglq44GTOA55WcWrzO8oFLTf_CtUiMk
fWe4qz9hfXnwZRO4jyBe2jgxWzTCsC3O1hh7hvJfCrE.zqngmHPWrG6hTglq44GTOA55WcWrzO8oFLTf_CtUiMk
```
but still similar error -
```
PS C:\WINDOWS\system32> (Get-PAOrder | Get-PAAuthorizations).challenges.error | fl


type   : urn:ietf:params:acme:error:unauthorized
detail : Incorrect TXT record "fWe4qz9hfXnwZRO4jyBe2jgxWzTCsC3O1hh7hvJfCrE.zqngmHPWrG6hTglq44GTOA55WcWrzO8oFLTf_CtUiMk" (and 1 more) 
         found at _acme-challenge.devcorp.mydomain.com
status : 403

type   : urn:ietf:params:acme:error:unauthorized
detail : Incorrect TXT record "fWe4qz9hfXnwZRO4jyBe2jgxWzTCsC3O1hh7hvJfCrE.zqngmHPWrG6hTglq44GTOA55WcWrzO8oFLTf_CtUiMk" (and 1 more) 
         found at _acme-challenge.devcorp.mydomain.com
status : 403

```

any other suggestion ?

**Comment by rmbolger** [comment]
Hmm...that's odd. Can you post the output of `nslookup -q=txt _acme-challenge.devcorp.mydomain.com.` *(don't forget the ending '.')* before you do the `Send-ChallengeAck`? I'm curious if there's something funky with how your DNS server is returning the results like extra quotes or both values on one line sort of thing. My output for two values looks something like this:

```
Server:  UnKnown
Address:  1.1.1.1

_acme-challenge.example.com     text =

        "asdf"
_acme-challenge.example.com     text =

        "qwer"
```

You could also try with just a single name rather than the two just for testing.

**Comment by chandan1001** [comment]
I did some more research with documents available for ACMEv2 and it turns out like key authorization for DNS has a few more steps than HTTP --

authToken generated by this code worked -

keyAuthHttp01 is value that is returned by  (Get-PAOrder | Get-PAAuthorizations).DNS01TOKEN |Get-KeyAuthorization



```
       var sha = SHA256.Create();
       var keyAuthzDig = sha.ComputeHash(Encoding.UTF8.GetBytes(keyAuthHttp01));
       var keyAuthDNS01=( UrlEncode(keyAuthzDig));


        public static string UrlEncode(byte[] raw)
        {
            string enc = Convert.ToBase64String(raw);  
            enc = enc.Split('=')[0];                  
            enc = enc.Replace('+', '-');              
            enc = enc.Replace('/', '_');
           return enc;
        }

```

**Comment by rmbolger** [comment]
You're totally right. I completely skipped a step for the manual DNS tutorial. The code you have there is basically taking a SHA256 hash of that KeyAuthorization value and encoding it as "Base64Url" which is a minor variation of Base64. Normally, `Publish-DNSChallenge` takes care of that for you here:

https://github.com/rmbolger/Posh-ACME/blob/fbda9ba8aea5c2de49662376bbda9d052b9c41bf/Posh-ACME/Public/Publish-DNSChallenge.ps1#L23-L29

But since my `ConvertTo-Base64Url` function is currently private, you'd have to do it manually like in your example code and futz with the SHA256 stuff yourself. I'm thinking I should add a switch like `-ForDns` to `Get-KeyAuthorization` that does it all for you. In the mean time, I'll update the wiki page with the missing steps.

**Comment by rmbolger** [comment]
The wiki page has been updated.

https://github.com/rmbolger/Posh-ACME/wiki/(Advanced)-Semi-Manual-DNS-Challenge-Validation#publishing-a-dns-challenge

**Comment by rmbolger** [comment]
And commit e220d6d has added `-ForDNS` to `Get-KeyAuthorization`. Feel free to test with the dev copy from the master branch. I'm not exactly sure when I'll cut a new release yet.
